### ISSUE TITLE
Typing

### ISSUE THREAD

**Issue by lambdaofgod** [stanfordnlp/dsp#83]
First of all, thanks for opensourcing your work, I think it's a great step in the direction of making LLM tools more sane.

I see that DSP uses dotdict which leads to extremely dynamic behavior that could lead to hard to debug errors when something will change between versions (field gets renamed, a required field is added et c), and is also problematic when someone would like to provide a HTTP API that will return your classes (because the developer needs to check which fields got set in code)

Questions:
- Do you plan to migrate the project to Pydantic/dataclasses that would validate arguments?
- If not, what are the obstacles?
- If you plan to migrate it but just don't have enough man-hours, could you point me to the classes that you think will have a stable API? I can migrate these classes to something that will support validation

I think I'll be forking your project in the coming weeks, it'll be great if I got some guidance so that my extension could be used by someone else.
Additional benefit of using pydantic/dataclasses would be that it'll be extremely easy to write a HTTP server (which is something that I'm very likely do anyways)

**Comment by okhat** [comment]
Hey, this sounds excellent and I want to dig deeper. Not sure if you follow my Twitter updates on this, but we have built a v2 (called DSPy) that we're releasing like ... in 2-3 days. It's not a rewrite actually, it's a much cleaner and more powerful wrapper around DSPv1, but I can't tell if it will resolve some of your concerns or not.

So I'd like to understand your concerns.

> dynamic behavior that could lead to hard to debug errors when something will change between versions

Versions of? The framework itself? Or a program built in it? Basically, I might need an example of this to understand what you mean.

What is the advantage of Pydantic here? What kind of validation are you thinking of?

**Comment by okhat** [comment]
One issue we did fix in DSPy (i.e., v2) is that the generate functions will take kwargs and not "x" (dsp.Example). This is a lot less opaque and causes fewer issues.

This sounds like it interacts with your concerns on validating which fields are where.

**Comment by lambdaofgod** [comment]
By validation I meant that for example when I removed 'demos' argument in `vanilla_LM_QA` from "demonstrate_search_predict" notebook I got an error in `generate` function and not in Example itself. Pydantic wouldn't let me create an object if I didn't pass the required field or if it had incorrect type (it would fail instantly thus making debugging far easier).

I'm looking forward for v2, when it drops I'll try to write these classes for validation.
